### PR TITLE
Implement repeating group widget mvp

### LINKED ISSUE #172: Implement repeating group widget mvp
-  [x] Include edit text as part of the repeating group widget
- [x] Create new root linear layout programatically. Add a unique id to this root repeating group linear layout  
- [x] Add repeating group layout definition as a nested value within the repeating group form element definition, so that we have the main edit text reference field followed by the repeating group layout, within the root linear layout
- [x] Add a static hash map of linear layout unique id to JSONArray representing the repeating group elements, which is parsed once when the repeating group is being set up
- [x] Add an edit text listener initially, and as a test (should add confirm button later), to the reference edit text in `attachLayout` or `getViewsFromJson` factory methods. 
- [x] On pressing `Enter` or `Done`, get the view id of root linear layout using `getParent` and use the id to add/remove views to/from the root layout
- [x] Call `getViewsFromJson` for each widget in the repeating group and add all elements returned from this function to the repeating group root layout. It should have all necessary validation added
- [x] To do this ^, parse each repeating group's element type, get factory for each type using `map.get(type)` in `JsonFormInteractor`, call the factory's `getViewFromJson` and pass the widget's JsonObject
- [x] Call the validate method in each widget's factory when validating elements in the repeating group (Edit: no need for this, validation is added for free when the above method is used to get widget layouts)
- [x]  Limit the number of repeating fields e.g. to 35 to avoid performance hits
- [x] Add a suffix of a unique string to field keys so it's easy to write back values to the form Json
- [x] Add relevance checks within the repeating group
- [x] Add a progress dialog as repeating group entries are being created
- [x] For deleted repeating group entries, the corresponding json object should be removed from the json form object
- [x] Add ability to specify reference edit text hint
- [x] Add ability to specify repeating group label text which will have a serially increasing suffix integer added to it
- [x] Add support for relative max validator

### PR DESCRIPTION
**Pull request by vincent-karuri**
Closes #172

### CODE REVIEW

File: android-json-form-wizard/src/main/java/com/vijay/jsonwizard/interactors/JsonFormInteractor.java [hunk 1]
```diff
@@ -44,7 +45,7 @@
 
     private static final String TAG = "JsonFormInteractor";
     private static final JsonFormInteractor INSTANCE = new JsonFormInteractor();
-    protected Map<String, FormWidgetFactory> map = new HashMap<>();
+    public static Map<String, FormWidgetFactory> map;
```
Comment (githengi) [inline android-json-form-wizard/src/main/java/com/vijay/jsonwizard/interactors/JsonFormInteractor.java]: should this be a static when its being modified in instance methods `registerWidgets()`

File: android-json-form-wizard/src/main/java/com/vijay/jsonwizard/widgets/RepeatingGroupFactory.java [hunk 2]
```diff
@@ -0,0 +1,327 @@
...
+
+    protected int MAX_NUM_REPEATING_GROUPS = 35;
+
+    private final ViewGroup.LayoutParams WIDTH_MATCH_PARENT_HEIGHT_WRAP_CONTENT = new LinearLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT);
+    private static Map<Integer, String> repeatingGroupLayouts = new HashMap<>();
+
+    private final String REFERENCE_EDIT_TEXT_HINT = "reference_edit_text_hint";
+    private final String REPEATING_GROUP_LABEL = "repeating_group_label";
+
+    protected final float REPEATING_GROUP_LABEL_TEXT_SIZE = 20;
```
Comment (githengi) [inline android-json-form-wizard/src/main/java/com/vijay/jsonwizard/widgets/RepeatingGroupFactory.java]: Consider using dimes resource to define this

File: android-json-form-wizard/src/main/java/com/vijay/jsonwizard/widgets/RepeatingGroupFactory.java [hunk 3]
```diff
@@ -0,0 +1,327 @@
...
+            attachRepeatingGroup(textView.getParent().getParent(), Integer.parseInt(textView.getText().toString()), widgetArgs);
+        } catch (Exception e) {
+            Log.e(TAG, e.getStackTrace().toString());
+        }
+    }
+
+    private void attachRepeatingGroup(final ViewParent parent, final int numRepeatingGroups, final WidgetArgs widgetArgs) {
+
+        if (numRepeatingGroups > MAX_NUM_REPEATING_GROUPS) {
+            return;
```
Comment (githengi) [inline android-json-form-wizard/src/main/java/com/vijay/jsonwizard/widgets/RepeatingGroupFactory.java]: Consider returning a validation error or some sort of hint to user what the problem is

File: android-json-form-wizard/src/main/java/com/vijay/jsonwizard/widgets/RepeatingGroupFactory.java [hunk 4]
```diff
@@ -0,0 +1,327 @@
...
+            @Override
+            protected void onPreExecute() {
+                showProgressDialog(R.string.creating_repeating_group_title, R.string.creating_repeating_group_message, widgetArgs.getFormFragment().getActivity());
+            }
+
+            @Override
+            protected List<View> doInBackground(Void... objects) {
+                int currNumRepeatingGroups = ((ViewGroup) parent).getChildCount() - 1;
+                diff = numRepeatingGroups - currNumRepeatingGroups;
+                for (int i = 0; i < diff; i++) {
```
Comment (githengi) [inline android-json-form-wizard/src/main/java/com/vijay/jsonwizard/widgets/RepeatingGroupFactory.java]: This caters for adding new repeating groups, consider informing user that incase they reduce number of repeating groups that those will not be reflected on form

File: android-json-form-wizard/src/main/java/com/vijay/jsonwizard/widgets/RepeatingGroupFactory.java [hunk 5]
```diff
@@ -0,0 +1,327 @@
...
+        repeatingGroup.addView(repeatingGroupLabel);
+
+        JSONArray repeatingGroupJson = new JSONArray(repeatingGroupLayouts.get(((LinearLayout) parent).getId()));
+        String uniqueId = UUID.randomUUID().toString();
+        for (int i = 0; i < repeatingGroupJson.length(); i++) {
+            JSONObject element = repeatingGroupJson.getJSONObject(i);
+            String elementType = element.optString(TYPE, null);
+            if (elementType != null) {
+                addUniqueIdentifiers(element, uniqueId);
+                FormWidgetFactory factory = JsonFormInteractor.getInstance().map.get(elementType);
```
Comment (githengi) [inline android-json-form-wizard/src/main/java/com/vijay/jsonwizard/widgets/RepeatingGroupFactory.java]: Consider getting JsonFormInteractor from FormFragment getPresenter which has the reference to interactor.

With approach then if a project overrides `JsonFormInteractor` which is normally the case then `JsonFormInteractor.getInstance() `will always the parent `JsonFormInteractor` and thereby skipping custom added/overridden factories

File: android-json-form-wizard/src/main/java/com/vijay/jsonwizard/widgets/RepeatingGroupFactory.java (hunk 3 above)
Comment (vincent-karuri) [reply android-json-form-wizard/src/main/java/com/vijay/jsonwizard/widgets/RepeatingGroupFactory.java]: This already exists.

File: android-json-form-wizard/src/main/java/com/vijay/jsonwizard/interactors/JsonFormInteractor.java (hunk 1 above)
Comment (vincent-karuri) [reply android-json-form-wizard/src/main/java/com/vijay/jsonwizard/interactors/JsonFormInteractor.java]: Not sure what the issue would be in this case but I've removed the static modifier.

File: android-json-form-wizard/src/main/java/com/vijay/jsonwizard/widgets/RepeatingGroupFactory.java (hunk 5 above)
Comment (vincent-karuri) [reply android-json-form-wizard/src/main/java/com/vijay/jsonwizard/widgets/RepeatingGroupFactory.java]: Done.

File: android-json-form-wizard/src/main/java/com/vijay/jsonwizard/widgets/RepeatingGroupFactory.java (hunk 2 above)
Comment (vincent-karuri) [reply android-json-form-wizard/src/main/java/com/vijay/jsonwizard/widgets/RepeatingGroupFactory.java]: Done.

**Pull request merged**